File: ts-RayTracer/src/sphere.ts
import { Hittable, HitRecord } from './hittable';
import { Interval } from './interval';
import { Ray } from './ray';
import { Point3, Vec3, sub, dot, mulVecScalar } from './vec3';

export class Sphere implements Hittable {
  constructor(private center: Point3, private radius: number, private mat: any) {
    this.radius = Math.max(0, radius);
  }

  hit(r: Ray, rayT: Interval, rec: HitRecord): boolean {
    const oc = sub(this.center, r.origin());
    const a = r.direction().lengthSquared();
    const h = dot(r.direction(), oc);
    const c = oc.lengthSquared() - this.radius * this.radius;

    const discriminant = h * h - a * c;
    if (discriminant < 0) return false;
    const sqrtd = Math.sqrt(discriminant);

    let root = (h - sqrtd) / a;
    if (!rayT.surrounds(root)) {
      root = (h + sqrtd) / a;
      if (!rayT.surrounds(root)) return false;
    }

    rec.t = root;
    rec.p = r.at(rec.t);
    const outwardNormal = divByScalar(sub(rec.p, this.center), this.radius);
    rec.setFaceNormal(r, outwardNormal);
    rec.mat = this.mat;
    return true;
  }
}

function divByScalar(v: Vec3, t: number): Vec3 {
  return mulVecScalar(v, 1 / t);
}

File: rust-Raytracer/src/sphere.rs
use crate::hittable::{HitRecord, Hittable};
use crate::interval::Interval;
use crate::material::Material;
use crate::ray::Ray;
use crate::vec3::{Point3, Vec3};

pub struct Sphere<M: Material> {
    pub center: Point3,
    pub radius: f64,
    pub mat: M,
}

impl<M: Material> Sphere<M> {
    pub fn new(center: Point3, radius: f64, mat: M) -> Self {
        Self {
            center,
            radius: radius.max(0.0),
            mat,
        }
    }
}

impl<M: Material> Hittable for Sphere<M> {
    fn hit<'a>(&'a self, r: &Ray, ray_t: Interval, rec: &mut HitRecord<'a>) -> bool {
        let oc = self.center - r.origin();
        let a = r.direction().length_squared();
        let h = Vec3::dot(r.direction(), oc);
        let c = oc.length_squared() - self.radius * self.radius;
        let discriminant = h * h - a * c;
        if discriminant < 0.0 {
            return false;
        }
        let sqrtd = discriminant.sqrt();
        let mut root = (h - sqrtd) / a;
        if !ray_t.surrounds(root) {
            root = (h + sqrtd) / a;
            if !ray_t.surrounds(root) {
                return false;
            }
        }
        rec.t = root;
        rec.p = r.at(rec.t);
        let outward_normal = (rec.p - self.center) * (1.0 / self.radius);
        rec.set_face_normal(r, outward_normal);
        rec.mat = Some(&self.mat);
        true
    }
}
